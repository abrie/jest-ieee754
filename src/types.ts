/* eslint-disable import/no-extraneous-dependencies */
/* tslint:disable:no-implicit-dependencies */
/* Because this is Jest extension, and Jest is present. */

import { MatcherState } from 'expect';
import { Context } from 'jest-snapshot';

/* SnapshotState interface is made to match https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/jest-snapshot/src/State.ts#L54
 */
export interface SnapshotState {
  _counters: Map<string, number>;
  _updateSnapshot: 'new' | 'all' | 'none';
  _snapshotData: Record<string, string>;
  _snapshotPath: string;
  markSnapshotsAsCheckedForTest: (testName: string) => void;
  _addSnapshot: (
    key: string,
    receivedSerialized: string,
    options: { isInline: boolean; error?: Error }
  ) => void;
  updated: number;
  added: number;
  unmatched: number;
  matched: number;
}

export SnapshotContext = Context

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeCloseToNumber(a: Number): CustomMatcherResult;
      toBeCloseToArraySnapshot(): CustomMatcherResult;
    }
  }
}
