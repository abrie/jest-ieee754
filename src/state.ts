/* eslint-disable no-underscore-dangle */
/* Because Jest StateSnapshot uses them. */

/* eslint-disable import/no-extraneous-dependencies */
/* tslint:disable:no-implicit-dependencies */
/* Because this is Jest extension, and Jest is present. */

import * as fs from 'fs';
import { utils } from 'jest-snapshot';
import { Context, SnapshotState } from './types';

export default class State {
  private currentTestName: string;
  private snapshotState: SnapshotState;
  private error?: Error;
  private hasSnapshot: boolean;

  private incrementCounters() {
    this.snapshotState._counters.set(
      this.currentTestName,
      (this.snapshotState._counters.get(this.currentTestName) || 0) + 1
    );
  }

  get count() {
    return Number(this.snapshotState._counters.get(this.currentTestName));
  }

  get key() {
    return utils.testNameToKey(this.currentTestName, this.count);
  }

  markSnapshotsAsCheckedForTest() {
    this.snapshotState.markSnapshotsAsCheckedForTest(this.currentTestName);
  }

  get snapshotIsPersisted() {
    return fs.existsSync(this.snapshotState._snapshotPath);
  }

  constructor(context: Context) {
    this.currentTestName = context.currentTestName;
    this.snapshotState = context.snapshotState;
    this.error = context.error;

    this.incrementCounters();

    this.hasSnapshot = this.getSnapshot() !== undefined;
  }

  getSnapshot() {
    return this.snapshotState._snapshotData[this.key];
  }

  setSnapshot(data: string) {
    this.snapshotState._snapshotData[this.key] = data;
  }

  /* This nested mess is derived the Jest snapshot matcher code:
   * https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/jest-snapshot/src/State.ts
   *
   * Which describes the conditions as follows:
   *  These are the conditions on when to write snapshots:
   * There's no snapshot file in a non-CI environment.
   * There is a snapshot file and we decided to update the snapshot.
   * There is a snapshot file, but it doesn't have this snaphsot.
   These are the conditions on when not to write snapshots:
   * The update flag is set to 'none'.
   * There's no snapshot file or a file without this snapshot on a CI environment.
   */
  couldAddSnapshot(): boolean {
    return (
      (this.hasSnapshot && this.snapshotState._updateSnapshot === 'all') ||
      ((!this.hasSnapshot || !this.snapshotIsPersisted) &&
        (this.snapshotState._updateSnapshot === 'new' ||
          this.snapshotState._updateSnapshot === 'all'))
    );
  }

  updateTally(pass: boolean) {
    if (this.couldAddSnapshot()) {
      if (this.snapshotState._updateSnapshot === 'all') {
        if (!pass) {
          if (this.hasSnapshot) {
            this.snapshotState.updated++;
          } else {
            this.snapshotState.added++;
          }
        } else {
          this.snapshotState.matched++;
        }
      } else {
        this.snapshotState.added++;
      }
    } else if (!pass) {
      this.snapshotState.unmatched++;
    } else {
      this.snapshotState.matched++;
    }
  }

  addSnapshot(data: string) {
    const isInline = false;

    this.snapshotState._addSnapshot(this.key, data, {
      error: this.error,
      isInline,
    });
  }
}
