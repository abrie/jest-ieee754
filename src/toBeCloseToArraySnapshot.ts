import { ulpDistance } from '@eirba/ieee754';
import { Context } from './types';
import State from './state';

function serialize(arr: number[]): string {
  return JSON.stringify(arr);
}

/* This function will throw an exception if param is not defined. */
function deserialize(str: string): number[] {
  return JSON.parse(str);
}

/* Catch failed deserialization and return an empty array to force a new snapshot */
function tryDeserialize(str: string): number[] {
  try {
    return deserialize(str);
  } catch {
    return [];
  }
}

function compare(want: number[], got: number[], maxUlp: bigint): boolean {
  if (want.length !== got.length) {
    return false;
  }

  return want.every((_, idx) => ulpDistance(want[idx], got[idx]) <= maxUlp);
}

export default function toBeCloseToArraySnapshot(
  this: Context,
  received: number[],
  maxUlp: bigint = BigInt(4)
) {
  const state = new State(this);

  const snapshot = state.getSnapshot();
  const expected = tryDeserialize(snapshot);

  const pass = compare(expected, received, maxUlp);

  state.markSnapshotsAsCheckedForTest();

  if (pass) {
    state.setSnapshot(serialize(received));
  }

  state.updateTally(pass);

  if (state.couldAddSnapshot()) {
    state.addSnapshot(serialize(received));

    return {
      message: () => '',
      pass: true,
    };
  }

  if (!pass) {
    return {
      message: () => `expected: ${expected}\n received: ${received}`,
      actual: serialize(received),
      count: state.count,
      expected: serialize(expected),
      key: state.key,
      pass: false,
    };
  }

  return {
    message: () => '',
    actual: serialize(received),
    count: state.count,
    expected: '',
    key: state.key,
    pass: true,
  };
}
