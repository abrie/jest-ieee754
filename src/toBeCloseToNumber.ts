import { ulpDistance } from '@eirba/ieee754';

export default function toBeCloseToNumber(got: number, want: number) {
  const diff = ulpDistance(want, got);
  const maxUlp = BigInt(4);
  const pass = diff <= maxUlp;

  if (pass) {
    return {
      message: () =>
        `received (${got}) and expected (${want}) differ by ${diff} ULP`,
      pass: true,
    };
  }
  return {
    message: () =>
      `received (${got}) and expected (${want}) differ by ${diff} ULP`,
    pass: false,
  };
}
