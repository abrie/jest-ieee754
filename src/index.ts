import toBeCloseToArraySnapshot from './toBeCloseToArraySnapshot'
import toBeCloseToNumber from './toBeCloseToNumber'

export { toBeCloseToArraySnapshot, toBeCloseToNumber }
