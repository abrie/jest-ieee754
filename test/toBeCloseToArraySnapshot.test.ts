import { toBeCloseToArraySnapshot } from '../src';

expect.extend({ toBeCloseToArraySnapshot });

describe('extension-test', () => {
  it('matches snapshot of literal array', () => {
    expect([1.234, 6.5542, 1.231, 999.99999]).toBeCloseToArraySnapshot();
  });

  it('matches snapshot of computed array', () => {
    const computed = [...Array(10)].map((_, idx) => Math.PI * idx);
    expect(computed).toBeCloseToArraySnapshot();
  });
});
