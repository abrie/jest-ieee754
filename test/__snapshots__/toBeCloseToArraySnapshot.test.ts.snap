// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension-test matches snapshot of computed array 1`] = `[0,3.141592653589793,6.283185307179586,9.42477796076938,12.566370614359172,15.707963267948966,18.84955592153876,21.991148575128552,25.132741228718345,28.274333882308138]`;

exports[`extension-test matches snapshot of literal array 1`] = `[1.234,6.5542,1.231,999.99999]`;
